### ISSUE TITLE
[NEW FEATURE] WebAssembly scripting system

### ISSUE THREAD

**Issue by MarkMcCaskey** [amethyst/amethyst#1729]
Hello! I'm a big fan of the Amethyst project, unfortunately I haven't had the opportunity to contribute yet.

I work at a company called [Wasmer](https://wasmer.io) and we're making a [server-side WebAssembly runtime](https://github.com/wasmerio/wasmer).  One of the neat uses of this is as a plugin system, being able to call WebAssembly functions from Rust and other languages and have the WebAssembly module call back into the host.

## Description

This is less of a feature proposal and more of the start of a discussion around using WebAssembly as a scripting language.  It's our goal to have Wasmer be very easy to use for these cases, but the Wasm ecosystem is still early on.  Whether using Wasmer or not, Wasm plugins are a neat idea that I think will become very popular over the coming years.

## What problem does this solve? What need does it fill?

Reasons why Wasm is useful as a scripting language:
- performance - Wasm JITs currently can run at near native speeds and we expect this to get better over time
- flexibility - use any language that compiles to Wasm; this is great for plugins created by users.  It also opens up tons of libraries and the tooling of all these languages
- security - Wasm can be sandboxed and metered.  It's simple to run untrusted code with limited resources and without worrying about it doing bad things (as of writing, metering works, but is still in development in Wasmer)
- standard - by using Wasm, you benefit from all the tools and work done on Wasm
- good plugin support - Wasm for plugins will likely become very popular and the tooling for connecting the pieces is coming together

## Describe the solution you'd like

Feedback on the idea!  Ideas about how to improve, what we'd need to support game engines well (we have some ideas already).

If you all are interested, we're interested in working with you to help design and implement this.

## Describe alternatives you've considered

- special, endorsed scripting language(s) - probably more work, less flexible, less secure, etc.
- JVM - licensing issues with embedding, less ecosystem support, less flexible (Wasm explicitly supports running without garbage collection)
- no scripting languages - user-created content is awesome + dynamism is nice to have during development
- V8/other JS Wasm VMs - limited by the needs of JS, likely not as focused on plugin use-cases as Wasmer (Wasmer has being easily embeddable as a core design goal)
- Other non-JS Wasm VMs - You shouldn't trust what I say here because I'm clearly biased, but Wasmer is pretty good and has some useful features and is set up in such a way (technically and organizationally) that I think we'd be a good fit.  Also if you use Wasmer then I'd be able to work on this during work 😄. Please do your own research here though and feel free to bring up any concerns!

## Additional context

Known issues:
- compilation: eventually Wasmer will get tiering allowing for more optimized code to be generated over time, but currently you have to commit to a level of optimization up-front
- in-development-ness of Wasm - we'll have to wait a bit for threads, built-in GC, SIMD in Wasm, and things like that

PS:
I wrote this before seeing https://github.com/amethyst/amethyst/issues/271 ; Wasm support could be built on top of this, but I think a Wasm engine as the driver of a scripting system will be in-practice much simpler if safety, flexibility, and performance are the goals.

**Comment by distransient** [comment]
Hello! Your interest in this is awesome. The issue you really want to be checking out is not that one, but https://github.com/amethyst/rfcs/pull/1. Going to tag @Moxinilian into this thread who has both authored that proposed interface but also has been the dominant force pushing it.

**Comment by Moxinilian** [comment]
WebAssembly has been discussed a lot in the context of the discussion around the scripting RFC. One of the reason we did not want wasm as the very core of the proposal is that it does not support very important languages for us (LuaJIT for example).

However, the interests of wasm are real, especially for modding support: one thing you did not mention for example is that Apple actually allows downloaded JITed scripts on iOS if they run in the version of WebKit they provide, which I assume has some form of wasm support. It also has all the safety guarantees that allow us to run untrusted scripts from third-party modders.

The idea we settled on regarding wasm is that this technology could be implemented as a sort of language driver sublayer, like another layer to build language drivers on to of. We could have LuaJIT be a language driver, and wasm+Rust be another one that would have some of the benefits of the Rust driver with additional safety, at the cost of the support for some of the ecosystem.

**Comment by MarkMcCaskey** [comment]
@Moxinilian 

I think that's a good place to start!  Wasm as a scripting system is new and there may be issues we're not aware of yet and I don't want to stop your existing work!  I think there are a lot of potential benefits to using Wasm widely and I'd like to see how well they survive contact with reality (I'm very optimistic!).

Yeah, I'm interested in learning more about the ecosystem support for LuaJIT (I'm reading this as features of the Amethyst ecosystem) and see how well we can do similar things with Wasm.

That's a good point about portability!

Something else I didn't mention is that Wasm could allow for even finer grained security; some of the new Wasm features (and a lot of the discussion around WASI) should make it possible to have an untrusted third-party script run untrusted "fourth-party" scripts in a way that's safe to everyone.  That level of security could have some very neat uses (for example, third-party mods executing code from the server or something).  The only downside to something like that from a game-engine perspective is that each Wasm module in that model would be running in its own linear memory, so if the execution crosses a lot of boundaries and touches a lot of memory, then it could have potentially really bad cache locality.  But I'm sure we can make things work!

Another neat feature of Wasm is that Wasm will be able to interact with Wasm in the same way that the host can interact with Wasm.  So it will be possible to port Wasm programs targeting another engine's interface/ABI to Amethyst through a Wasm-only layer.  Mitigating vendor lockin is a huge win for FOSS!

**Comment by erlend-sh** [comment]
> Another neat feature of Wasm is that Wasm will be able to interact with Wasm in the same way that the host can interact with Wasm. So it will be possible to port Wasm programs targeting another engine's interface/ABI to Amethyst through a Wasm-only layer. Mitigating vendor lockin is a huge win for FOSS!

That’s a very cool prospect indeed! I don’t suppose there’s already some real world examples of this?

**Comment by MarkMcCaskey** [comment]
@erlend-sh

Most Wasm engines don't support dynamic linking yet as far as I know, but I don't think there's any major blocking issues for it, just uncertainty about how we want to standardize.  There's a lot of excitement about "shared-nothing" linking as a possible default (introduced in the slides below).  As I understand it, the Wasm module as compatibility layer is the current plan for how WASI will handle backwards compatibility, but nothing is set in stone yet.

https://docs.google.com/presentation/d/1wtAknL-UJWDoIgSbyF5paTBSpVVj-fKU4tiHMxJbSzE/edit#slide=id.g5b3b90bd10_0_0

is a fantastic set of slides that talks about linking types and has useful diagrams for how Wasm modules would fit together (for example, exports from A connect to imports from B)

**Comment by OvermindDL1** [comment]
> lua

https://github.com/vvanders/wasm_lua

For note, there are a lot of languages already on wasm, including lua (though not JIT'd, but still).

**Comment by fhaynes** [comment]
@MarkMcCaskey how do you see multiple scripts functioning? Would each script run in a dedicated WASM VM? Would we keep a pool of WASM VMs around, and if so, how do we sanitize each before re-use? Or could one VM run multiple scripts?

**Comment by MarkMcCaskey** [comment]
@fhaynes with wasmer you can have many 'instance's (should be on the order of thousands) running at once; they're cheap to create and start.

It depends on the use case.  You could probably get better performance by reusing pieces of instances, but I don't believe wasmer has any APIs for that right now though.  If you're executing entire, new scripts in a tight loop that could be a bottleneck.  I think the general use case would be back and forth communication though.  You'd start a script once and then on every update or render or misc event you'd call into it and it would do its thing (calling back into the engine to have side-effects or to store data) and then return (when it finishes or if the host doesn't want to wait anymore).

If the scripts are just interfacing with Amethyst, then that should be sufficient.  If the scripts need to be able to pass data between each other directly or trigger execution of code in other instances, then we'd have to do one of the linking styles mentioned in the slides that I linked in a previous comment.

Those linking ideas are new and we haven't started working on them yet, but we're happy to prioritize features that people need!

**Comment by fhaynes** [comment]
@MarkMcCaskey Hmm, nifty. Forgive me, I've not yet made it through your repo, though I saw an architecture document. Do ya'll have a Slack/Discord/IRC/carrier pigeon network I can come ask annoying questions on?

**Comment by syrusakbary** [comment]
Hi @fhaynes I can invite you to a single channel in our Slack org.
What's your email?

**Comment by fhaynes** [comment]
@syrusakbary fletcher@amethyst-engine.org

**Comment by zicklag** [comment]
I am working on the [Arsenal](https://github.com/katharostech/arsenal) project which seeks to provide a Blender integration with Amethyst. To improve the user experience, we use an Amethyst core that is pre-compiled for users so that they never have to compile Amethyst to start game development. This means that *all* of the game code will have to interact with the engine through the scripting interface.

WASM seems ideal for this in almost every way, but my concern is that, as far as I understand, JIT compilers can't run on game consoles. This would make put Wasmer at a performance disadvantage when deploying to consoles, compared to using a pure C API. Do I understand that correctly, or is there any way around that?

This is probably completely impossible, but is there any way you could ahead-of-time compile the WASM modules when deploying to consoles, and still use an Amethyst WASM API for those modules?

**Comment by fhaynes** [comment]
> This is probably completely impossible, but is there any way you could ahead-of-time compile the WASM modules when deploying to consoles, and still use an Amethyst WASM API for those modules?

@zicklag 

I'm only beginning to experiment with wasmer this weekend, so it is far from a sure thing. And even if we do integrate it, it would not preclude us from offering a pure C API, which I think we should do regardless. Too many things expect you to have one for integration. Wasmer (or similar) is more to allow people to use a scripting language.

I'm not super familiar with how you are integrating with Blender. Can you give me more details or a link so I can take a look?

**Comment by zicklag** [comment]
>  it would not preclude us from offering a pure C API, which I think we should do regardless. 

OK, that is good. As long as we have a C API, then we can still write Rust "scripts" without any performance hit.

You can see more of how we are integrating with Blender [here](https://katharostech.github.io/arsenal/architecture/draft-2.html). Scripting itself will not be very integrated with Blender as you will probably write code in a dedicated IDE, but the major point is that all of the game code has to be able to be written without having to re-compile Amethyst, because Amethyst will be embedded into the Arsenal Blender plugin and won't change during game development.

**Comment by MarkMcCaskey** [comment]
> I am working on the [Arsenal](https://github.com/katharostech/arsenal) project which seeks to provide a Blender integration with Amethyst. To improve the user experience, we use an Amethyst core that is pre-compiled for users so that they never have to compile Amethyst to start game development. This means that _all_ of the game code will have to interact with the engine through the scripting interface.
> 
> WASM seems ideal for this in almost every way, but my concern is that, as far as I understand, JIT compilers can't run on game consoles. This would make put Wasmer at a performance disadvantage when deploying to consoles, compared to using a pure C API. Do I understand that correctly, or is there any way around that?
> 
> This is probably completely impossible, but is there any way you could ahead-of-time compile the WASM modules when deploying to consoles, and still use an Amethyst WASM API for those modules?

That's really neat!  Blender is something I've been thinking about too and I think Wasm could be a great fit there, allowing for game engine code and blender plugins to share code could make things easier.

Wasmer is AOT compiled, not really JITed, we're adding more JIT-like features, but it will still be usable as a JIT.  There's another Wasm runtime called Lucet which compiles ahead of time into Linux executables with the runtime embedded in that executable.  I'm under the impression that doing a similar thing with Wasmer would not be particularly difficult, we just haven't had a use case for something like that yet.  So precompiling Wasm into native binaries/libraries for restricted platforms or avoiding the one time compilation cost of static modules in a released version is certainly possible!

**Comment by zicklag** [comment]
> That's really neat! Blender is something I've been thinking about too and I think Wasm could be a great fit there, allowing for game engine code and blender plugins to share code could make things easier.

I hadn't thought of using WASM *in* Blender yet, but I'll keep my mind open to the possibility if the opportunity presents itself. We are currently writing the majority of the Blender plugin in Rust using PyO3, so we do already get to share code between the engine and the plugin, which is nice.

> So precompiling Wasm into native binaries/libraries for restricted platforms or avoiding the one time compilation cost of static modules in a released version is certainly possible!

That is really nice to hear. I think that is my only real concern with WASM for my use-case, other than Windows support, which is under development anyway. Modding is something very important to my team so having a modding interface that allows for all kinds of languages to be used with high performance *and* sandboxing, without having to build mods for each platform ( Windows, Mac, Linux, etc. ) is pretty much as good as it could get.

**Comment by erlend-sh** [comment]
Tangentially related: https://gitlab.com/veloren/veloren/issues/181

**Comment by zicklag** [comment]
@MarkMcCaskey In order to get any data from the Amethyst game into a WASM "script" that needs to operate on that data, you have to copy the data into the WASM VM's linear memory. Once the WASM script has processed the data it also needs to be copied back into the Amethyst game's memory, though maybe that is similar to what you would have to do with any other scripting language. ( If I'm understanding this right, I'm pretty new to some of this. )

Do you know how much of a performance hit that represents or if there are any particularly more efficient ways to handle that in general? If you were writing an entire game through the scripting interface, would it be much more performant to use a pure C API instead of the WASM interface? I know that WASM itself can execute at near native speeds, but even if the WASM code itself runs very fast, it is still copying memory back and forth between WASM and Rust.

For example, if I were to write Rust code for scripting Amethyst through a C API you could avoid the memory copying because you could pass pointers over the C API. If I were writing Rust for scripting Amethyst through the WASM API you would have to copy memory in and out of the WASM VM. That is partly how WASM achieves the security, though, because WASM can't arbitrarily access the memory in the Amethyst game. There isn't a way to pass references to memory in the Amethyst game to WASM, right? I don't think that would make sense.

Anyway, I'm mostly trying to make sure that I have an accurate picture in my head of how this works. I want to understand whether or not it would make sense to use a C API for building my "core game" and only use WASM for mods, or whether it would be perfectly fine to use WASM for the entire game.

**Comment by MarkMcCaskey** [comment]
@zicklag That's correct, the Wasm ("guest") memory is a subset of the host's memory.  The host can freely access and use the guest memory as it wants, but can only trust the data there as far as it can trust the Wasm module.  You are correct that this is overhead and that depending on the level of trust and the amount of data, this could be a big problem.

In the case of all Wasm code being written by you, then it's safe for the host to have a high level of trust.  In this case, the strongest optimization is that you store the data in the Wasm module's memory, which the host can freely interact with.  This essentially removes the sandbox, especially if the host derefs host pointers stored in that memory...

In the probably not too far off future when the interaction between multiple Wasm modules is better defined, it'll be possible for a Wasm module to interact with other Wasm modules with various levels of trust.  So it'll be possible to mix and match these styles.  There are definitely trade offs between security and performance and it depends a lot on the specifics of exactly what you're doing.

Off the top of my head, a good design for an API for use with completely untrusted code would be something in a query, mutate style.  So the host system would be acting like a database and passing back small views into the data (really chunks, because they're copied) that the guest can do logic on and then call exposed functions to apply mutations.

Passing data back and forth in a trustless context means that the host has to do a lot of copying and that the host has to do a lot of validation (all the data passed back has to be assumed to be malicious until validated), so minimizing data shared and controlling the changes that can be made (through explicit API calls (as opposed to allowing a mod to take in an entity as an object and make arbitrary changes and then the host verifying those changes and applying them)) is key, I think.

This limited view form also has the benefit of allowing better theoretical concurrency given a sufficiently advanced database/engine (a.k.a. magic).

This post is pretty vague and I'd be interested in having more in-depth discussions about these kinds of things sometime with anyone that's interested.

**Comment by Moxinilian** [comment]
To save on copying you can use fat pointer mechanisms similar to what would be done with LuaJIT. It covers most of the important Amethyst use cases.

**Comment by zicklag** [comment]
@MarkMcCaskey Thanks for the explanation.

I think for Arsenal I will want to use a C API for scripting the fully trusted core game, and then use WASM for scripting sandboxed mods. That way we get a zero cost API for the unsandboxed game, but we can still support sandboxed WASM mods that will have limited control so that they can be run safely.

It would be cool, too if we could design the scripting API so that it was the same, or nearly the same, for mods and the game so that you wouldn't have to change the API that you use when switching from one to the other.

**Comment by fhaynes** [comment]
Sorry everyone, I've been very busy. I'll weigh in on this as soon as I can, hopefully in the next day or two.

**Comment by ryzokuken** [comment]
@MarkMcCaskey can you drop me an invite to the wasmer Slack so we could discuss details regarding this one over there?

**Comment by erlend-sh** [comment]
PoC in progress in #1892

**Comment by MarkMcCaskey** [comment]
> @MarkMcCaskey can you drop me an invite to the wasmer Slack so we could discuss details regarding this one over there?

Sorry about the delay, thought we had sent you an invite but I don't think we did -- you should have one now!

**Comment by caelunshun** [comment]
I've found that `wasmer_runtime::Instance` is neither `Send` nor `Sync`—I this will prevent modules from being used from or as systems. Are there any plans to get around this?

**Comment by syrusakbary** [comment]
I believe that we can fix that easily. I created an issue in Wasmer to do it: https://github.com/wasmerio/wasmer/issues/748

**Comment by MarkMcCaskey** [comment]
So we're in the process of seriously investigating our architecture to make Instance usable for this case and for the future feature of threads inside of Wasm.

As part of this, I made a hello world specs example by unsafely marking `Instance` Send + Sync  (We use `Rc` internally and other things, so this definitely wrong):

In `wasmer/lib/runtime-core/src/instance.rs`:
```Rust
unsafe impl Send for Instance {}
unsafe impl Sync for Instance {}
```

In `wasmer-specs-example/src/main.rs`
```Rust
#[macro_use]
extern crate specs_derive;

use specs::prelude::*;
use wasmer_runtime::{func, imports, instantiate, Ctx, Instance, Value};
use wasmer_runtime_core::{memory::ptr::WasmPtr, types::ValueType};

#[derive(Debug, Component)]
#[storage(VecStorage)]
struct Vel(f32, f32);

#[derive(Debug, Component, Clone, Copy)]
#[storage(VecStorage)]
#[repr(C)]
struct Pos(f32, f32);

unsafe impl ValueType for Pos {}

#[derive(Debug, Component)]
#[storage(VecStorage)]
struct Name(String);

struct Move;

impl<'a> System<'a> for Move {
    type SystemData = (
        WriteExpect<'a, ScriptingSystem>,
        WriteStorage<'a, Pos>,
        ReadStorage<'a, Vel>,
    );

    fn run(&mut self, (mut ss, mut pos, vel): Self::SystemData) {
        let instance = &mut ss.instance;

        let update_position = instance
            .func::<(f32, f32, f32, f32), u64>("update_position")
            .unwrap();
        for (pos, vel) in (&mut pos, &vel).join() {
            let result: u64 = update_position.call(pos.0, pos.1, vel.0, vel.1).unwrap();
            *pos = unsafe { std::mem::transmute(result) };
        }
    }
}

struct ShoutOut;

impl<'a> System<'a> for ShoutOut {
    type SystemData = (ReadStorage<'a, Name>, ReadStorage<'a, Pos>);

    fn run(&mut self, (name, pos): Self::SystemData) {
        for (name, pos) in (&name, &pos).join() {
            println!("{} is at {}, {}!", name.0, pos.0, pos.1);
        }
    }
}

static WASM: &'static [u8] =
    include_bytes!("../../specs-script/target/wasm32-unknown-unknown/release/specs-script.wasm");

struct ScriptingSystem {
    instance: Instance,
}

fn main() {
    let import_object = imports! {};

    let instance = instantiate(WASM, &import_object).unwrap();

    let scripting_system = ScriptingSystem { instance };

    let mut world = World::new();
    world.add_resource(scripting_system);

    let mut dispatcher = DispatcherBuilder::new()
        .with(Move, "move", &[])
        .with(ShoutOut, "shout_out", &["move"])
        .build();

    dispatcher.setup(&mut world);

    world
        .create_entity()
        .with(Vel(2.0, 0.))
        .with(Pos(0.0, 0.))
        .with(Name("Ario".to_string()))
        .build();
    world
        .create_entity()
        .with(Vel(4.0, 1.))
        .with(Pos(1.6, 1.))
        .with(Name("Bario".to_string()))
        .build();
    world
        .create_entity()
        .with(Vel(1.5, 2.))
        .with(Pos(5.4, -2.))
        .with(Name("Cario".to_string()))
        .build();

    world.create_entity().with(Pos(2., 2.)).build();

    for _ in 0..30 {
        dispatcher.dispatch_par(&world);
        world.maintain();
    }
}
```

In `specs-script/src/main.rs`:
```Rust
#[repr(C)]
struct Pos(f32, f32);

#[no_mangle]
fn update_position(x: f32, y: f32, vel_x: f32, vel_y: f32) -> u64 {
    let pos = Pos(x + vel_x, y + vel_y);
    let ret: u64 = unsafe { std::mem::transmute(pos) };

    ret
}

fn main() {}
```

In `wasmer-specs-example/Cargo.toml`:
```toml
[package]
name = "wasmer-specs-example"
version = "0.1.0"
authors = ["Mark McCaskey"]
edition = "2018"

[dependencies]
specs = "0.15"
specs-derive = "0.4.0"
wasmer-runtime-core = { path = "../../wasmer/lib/runtime-core" }
wasmer-runtime = { path = "../../wasmer/lib/runtime" }
```

And with this, it works.

Here's what I got out of the example -- I'm hoping to get feedback from Amethyst and @arlyon / people at feather on some of these points:

1. There are a number of serious ergonomic issues this example shows, for example, returning or operating on more data becomes unpleasant and slow.  `WasmPtr` (not shown in this example) provides a safe and not too awful interface for getting data from pointers into the guest memory, but we'll clearly need more than that.

2. The difficultly of inter-operating with specs.  Specs is typesafe so having a dynamic system is not possible as far as I can tell.   Additionally, it doesn't seem obvious to me if it's reasonably possible to have scripts registering their own components.  And if we can't dynamically make systems, then there's no point in dynamic components anyways. This is a problem with all dynamic scripting of specs though.  My question to Amethyst and Feather is: is this interface sufficient for the types of scripting you want to support?  Looking for lots of information here, so please share as much as can!

3. Misc. There's a lot more work to be done in this example; we should show using a shared function cache, executing wasm scripts in parallel, and prevent the wasm script from blocking indefinitely.

Feel free to play with the example above and share your thoughts and ideas!

**Comment by OvermindDL1** [comment]
It wouldn't be too hard to make a trait that you can implement for the types you want to dynamically expose to scripting.  The trait can start off with a set of simple functions like `getCallables() -> VecOrSliceOrWhatever<Callables>` and `getCallable(name: &str) -> Option<Callable>`.  Or you could make an entire path accessing system like I made in my old C++ engine.  It makes accessing things via scripting very easy, and you can always make better methods for hot paths when necessary, but it at least makes it easy to expose anything.

**Comment by MarkMcCaskey** [comment]
@OvermindDL1 That makes sense to me, but I don't think specs itself exposes that kind of functionality -- all access patterns for systems are known at compile time.  To clarify and restate what you're saying: the solution is to have a side-table/database outside the ECS that can be used to get access to specific data in the ECS for scripting access.

If Wasmer isn't managed by specs, then it doesn't strictly need to be `Send + Sync`.  That would make it harder for certain use cases though, like a script subroutine that executes on every collision, a pattern which could be a nice way to do hot-swappable debugging (add log statements, draw squares around each collision, etc.)

Please correct me if I misunderstood you.

**Comment by OvermindDL1** [comment]
> but I don't think specs itself exposes that kind of functionality 

It does not, but it potentially could for the basic storage types, as long as the contained data type is also exposed.

I have a post somewhere either here on amethyst or on specs about the old `Node` system in my C++ engine that was for exposing everything needed to scripts.

Even if types implemented such a thing they shouldn't be default exposed to scripts, that should be controlled solely by the user of amethyst.  For example I had scripts for various purposes, GUI controlling ones had access to certain things, Mods had access to other things (some overlap), etc... etc...  I fine-tuned controlled what each script had access to (which also allowed me to parallize access further).

**Comment by erlend-sh** [comment]
Are the constraints basically the same with [Legion](https://github.com/TomGillen/legion) instead of specs? @TomGillen @kabergstrom @zicklag 

It is looking quite likely that we will eventually switch out specs for Legion, as per this discussion:

https://community.amethyst.rs/t/legion-ecs-discussion/965?u=erlend_sh

That’s still several months off at best though, so specs should still be considered the primary target.

**Comment by MarkMcCaskey** [comment]
@erlend-sh Thanks!  I don't have much context on Legion right now, I read over some of the chat in the link you posted and it seems that Legion doesn't support adding components to entities dynamically.  The question above is really a design question about the role of scripts and/or mods in Amethyst.  Additionally, I think any ECS designed to be extremely fast in Rust will, in general, not support dynamic operations that you might want from a scripting system.  The reason I think that is due to how Rust libraries make things fast and just the fundamental fact that optimization requires assumptions and being dynamic invalidates a lot of common assumptions.

There might be a better place to continue this conversation, but nothing on this topic is specific to using Wasm right now.

**Comment by caelunshun** [comment]
I've done some pondering about the Specs problem and proposed a solution [here](https://github.com/caelunshun/feather/issues/106#issuecomment-529080119). Any thoughts on this idea?

**Comment by khionu** [comment]
Still relevant

**Comment by erlend-sh** [comment]
Broader discussion is available on Rust Users: [WASM as a Platform for Abstraction](https://users.rust-lang.org/t/wasm-as-a-platform-for-abstraction/35736?u=erlend_sh).
